### PR TITLE
Adds race task

### PR DESCRIPTION
**Pull request by flovilmart**
Adds a race task, been using that code in a category in production apps for a while.
Heavily inspired by the `taskForCompletionOfAll:`

### CODE REVIEW

**Comment by nlutsenko** [comment]
Wonderful! Yes, this is helpful for everyone, and is surfaced on Android as `whenAny`.
Meaning that here it would be `taskForCompletionOfAnyTask:` :grin: 

Will review code in detail in a little bit, but the first thing - can we rename it to `taskForCompletionOfAnyTask:`.
Slightly a longer name, I know, but overall - it describes intention better.

**Comment by flovilmart** [comment]
`taskForCompletionOfAnyTask:` suggest that it requires completion involved. `race:` implies that the faster to return would be the one... I dunno, I like the conciseness of race (and I have it implemented in about 10 projects)

**Comment by flovilmart** [comment]
@nlutsenko adding more tests, support for nullable array and rename the method

**Comment by nlutsenko** [comment]
Going to nitpick on stuff here, but will try my best to review the entire thing right now.

File: Bolts/Common/BFTask.m [hunk 1]
```diff
@@ -172,6 +172,75 @@ + (instancetype)taskForCompletionOfAllTasksWithResults:(NSArray<BFTask *> *)task
...
+    NSMutableArray *exceptions = [NSMutableArray new];
+    
+    BFTaskCompletionSource *source = [BFTaskCompletionSource taskCompletionSource];
+    for (BFTask *task in tasks) {
+        [task continueWithBlock:^id(BFTask *task) {
+            if (task.exception != nil) {
+                @synchronized(lock) {
+                    [exceptions addObject:task.exception];
+                }
+            }else if (task.error != nil) {
```
Comment (nlutsenko) [inline Bolts/Common/BFTask.m]: Nit: Space before `else`.

File: Bolts/Common/BFTask.m [hunk 2]
```diff
@@ -172,6 +172,75 @@ + (instancetype)taskForCompletionOfAllTasksWithResults:(NSArray<BFTask *> *)task
...
+                @synchronized(lock) {
+                    [errors addObject:task.error];
+                }
+            }else if (task.cancelled) {
```
Comment (nlutsenko) [inline Bolts/Common/BFTask.m]: Nit: Space before `else`.

File: Bolts/Common/BFTask.m [hunk 3]
```diff
@@ -172,6 +172,75 @@ + (instancetype)taskForCompletionOfAllTasksWithResults:(NSArray<BFTask *> *)task
...
 }
 
++ (instancetype)taskForCompletionOfAnyTask:(NSArray<BFTask *> *)tasks
+{
+    __block int32_t total = (int32_t)tasks.count;
+    if (total == 0) {
+        return [self taskWithResult:nil];
+    }
+    
+    __block dispatch_once_t token = 0;
```
Comment (nlutsenko) [inline Bolts/Common/BFTask.m]: dispatch_once_t must point to a variable stored in a global or static scope. The result of using a predicate like this is actually `undefined`, according to documentation:
https://developer.apple.com/library/ios/documentation/Performance/Reference/GCD_libdispatch_Ref/

You can use `OSAtomicTestAndSet` for this with a boolean.

File: Bolts/Common/BFTask.m [hunk 4]
```diff
@@ -172,6 +172,75 @@ + (instancetype)taskForCompletionOfAllTasksWithResults:(NSArray<BFTask *> *)task
...
+                }
+            }else if (task.error != nil) {
+                @synchronized(lock) {
+                    [errors addObject:task.error];
+                }
+            }else if (task.cancelled) {
+                OSAtomicIncrement32(&cancelled);
+            }
+            
+            if (task.result != nil) {
```
Comment (nlutsenko) [inline Bolts/Common/BFTask.m]: Task can finish with a `nil` result, which should still mark the completion of this task to match closer with other methods here.

File: Bolts/Common/BFTask.m [hunk 5]
```diff
@@ -172,6 +172,75 @@ + (instancetype)taskForCompletionOfAllTasksWithResults:(NSArray<BFTask *> *)task
...
+            
+            if (OSAtomicDecrement32(&total) == 0 && token == 0) {
+                if (cancelled > 0) {
+                    [source cancel];
+                } else if (exceptions.count > 0) {
+                    if (exceptions.count == 1) {
+                        source.exception = [exceptions objectAtIndex:0];
+                    } else {
+                        NSException *exception =
+                        [NSException exceptionWithName:@"BFMultipleExceptionsException"
```
Comment (nlutsenko) [inline Bolts/Common/BFTask.m]: Use `BFTaskMultipleExceptionsException` instead of a string literal.

File: Bolts/Common/BFTask.m [hunk 6]
```diff
@@ -172,6 +172,75 @@ + (instancetype)taskForCompletionOfAllTasksWithResults:(NSArray<BFTask *> *)task
...
+                                                reason:@"There were multiple exceptions."
+                                              userInfo:@{ @"exceptions": exceptions }];
+                        source.exception = exception;
+                    }
+                } else if (errors.count > 0) {
+                    if (errors.count == 1) {
+                        source.error = [errors objectAtIndex:0];
+                    } else {
+                        NSError *error = [NSError errorWithDomain:@"bolts"
```
Comment (nlutsenko) [inline Bolts/Common/BFTask.m]: Use `BFTaskErrorDomain` instead of a string literal.

File: Bolts/Common/BFTask.m [hunk 7]
```diff
@@ -172,6 +172,75 @@ + (instancetype)taskForCompletionOfAllTasksWithResults:(NSArray<BFTask *> *)task
...
+                    } else {
+                        NSException *exception =
+                        [NSException exceptionWithName:@"BFMultipleExceptionsException"
+                                                reason:@"There were multiple exceptions."
+                                              userInfo:@{ @"exceptions": exceptions }];
+                        source.exception = exception;
+                    }
+                } else if (errors.count > 0) {
+                    if (errors.count == 1) {
+                        source.error = [errors objectAtIndex:0];
```
Comment (nlutsenko) [inline Bolts/Common/BFTask.m]: Nit: `[errors firstObject]`

File: Bolts/Common/BFTask.m [hunk 8]
```diff
@@ -172,6 +172,75 @@ + (instancetype)taskForCompletionOfAllTasksWithResults:(NSArray<BFTask *> *)task
...
+                        source.error = error;
+                    }
+                }
+            }
+            // Abort execution of per tasks continuations
+            return nil;
+        }];
+    }
+    
+
```
Comment (nlutsenko) [inline Bolts/Common/BFTask.m]: Nit: Remove whitespace at 238-239.

File: Bolts/Common/BFTask.h [hunk 9]
```diff
@@ -83,6 +83,14 @@ typedef __nullable id(^BFContinuationBlock)(BFTask<ResultType> *task);
 + (instancetype)taskForCompletionOfAllTasksWithResults:(nullable NSArray<BFTask *> *)tasks;
 
 /*!
+ Returns a task that will be completed once there is at least one successful task.
+ The first task to successuly complete will set the result, all other tasks results are 
+ ignored.
```
Comment (nlutsenko) [inline Bolts/Common/BFTask.h]: Maybe expand on behavior of errors/exceptions/cancellation, as without documentation it's an implementation detail at this point.

**Comment by nlutsenko** [comment]
Tried my best to review everything. There are few important pieces (usage of dispatch once predicate and constants instead of magic strings), but overall - it looks complete and nice.

Discussion piece:
Do you think it would be beneficial to have a `taskForCompletionOfAnyTask:cancellationToken:` in addition to this one? Most of our methods do have it, but I actually don't think it's that much helpful, plus it changes the semantics of how we treat cancellation, as we probably should cancel the whole task altogether if the token is cancelled.
Just throwing this here for discussion and curious to know your opinion on this.

File: Bolts/Common/BFTask.m (hunk 4 above)
Comment (flovilmart) [inline Bolts/Common/BFTask.m]: :+1:

File: Bolts/Common/BFTask.m (hunk 3 above)
Comment (flovilmart) [inline Bolts/Common/BFTask.m]: Good point, didn't knew about that.

**Comment by flovilmart** [comment]
for the cancellation token, I believe that can be problematic. It's heavily inspired by `+ (instancetype)taskForCompletionOfAllTasks:(nullable NSArray<BFTask *> *)tasks;` that don't have cancellation tokens, and I believe for the same reason. If tasks need a cancellation token, they should be cancellable by themselves don't you think?

**Comment by nlutsenko** [comment]
Cancellable by themselves - maybe, though I sort of can see a situation where you would want to have an ability to cancel the entire chain that ends with `AnyTask` if say a person quite from the view, but still send the updates to the server. Up to you if you want to build another method for it here (we probably would want 2 methods one with token one without to match other APIs).

File: Bolts/Common/BFTask.m [hunk 10]
```diff
@@ -174,6 +174,74 @@ + (instancetype)taskForCompletionOfAllTasksWithResults:(nullable NSArray<BFTask
...
+                }
+            } else if (task.error != nil) {
+                @synchronized(lock) {
+                    [errors addObject:task.error];
+                }
+            } else if (task.cancelled) {
+                OSAtomicIncrement32Barrier(&cancelled);
+            }
+            
+            if (task.completed && !task.faulted && !task.cancelled) {
```
Comment (nlutsenko) [inline Bolts/Common/BFTask.m]: I feel like you can unify this as `else` statement on the above condition. Also, there is no way this task is not going to be completed here, since `continueWithBlock:` is called only after it's completed.

File: Bolts/Common/BFTask.m [hunk 11]
```diff
@@ -174,6 +174,74 @@ + (instancetype)taskForCompletionOfAllTasksWithResults:(nullable NSArray<BFTask
...
 }
 
++ (instancetype)taskForCompletionOfAnyTask:(nullable NSArray<BFTask *> *)tasks
+{
+    __block int32_t total = (int32_t)tasks.count;
+    if (total == 0) {
+        return [self taskWithResult:nil];
+    }
+    
+    __block int succeeded = 0;
```
Comment (nlutsenko) [inline Bolts/Common/BFTask.m]: Name suggestion: `completed`, since it doesn't necessarily mark success.

File: Bolts/Common/BFTask.m [hunk 12]
```diff
@@ -174,6 +174,74 @@ + (instancetype)taskForCompletionOfAllTasksWithResults:(nullable NSArray<BFTask
...
+                    } else {
+                        NSException *exception =
+                        [NSException exceptionWithName:BFTaskMultipleExceptionsException
+                                                reason:@"There were multiple exceptions."
+                                              userInfo:@{ @"exceptions": exceptions }];
+                        source.exception = exception;
+                    }
+                } else if (errors.count > 0) {
+                    if (errors.count == 1) {
+                        source.error = [errors firstObject];
```
Comment (nlutsenko) [inline Bolts/Common/BFTask.m]: Nit: Use dot-syntax, since `firstObject` is a property.

File: Bolts/Common/BFTask.m [hunk 13]
```diff
@@ -174,6 +174,74 @@ + (instancetype)taskForCompletionOfAllTasksWithResults:(nullable NSArray<BFTask
...
+                }
+            }
+            
+            if (OSAtomicDecrement32Barrier(&total) == 0 &&
+                OSAtomicCompareAndSwap32Barrier(0, 1, &succeeded)) {
+                if (cancelled > 0) {
+                    [source cancel];
+                } else if (exceptions.count > 0) {
+                    if (exceptions.count == 1) {
+                        source.exception = [exceptions firstObject];
```
Comment (nlutsenko) [inline Bolts/Common/BFTask.m]: Nit: Use dot-syntax, since `firstObject` is a property.

File: BoltsTests/TaskTests.m [hunk 14]
```diff
@@ -628,6 +629,83 @@ - (void)testTaskForCompletionOfAllTasksWithResultsNoTasksImmediateCompletion {
...
+        return [BFTask taskWithResult:@"second"];
+    }];
+
+    
+    BFTask * task = [BFTask taskForCompletionOfAnyTask:@[first, second]];
+    [task waitUntilFinished];
+    
+    XCTAssertEqualObjects(@"first", task.result);
+}
+
```
Comment (nlutsenko) [inline BoltsTests/TaskTests.m]: Nit: Remove unneeded whitespace at 653/654.

File: BoltsTests/TaskTests.m [hunk 15]
```diff
@@ -628,6 +629,83 @@ - (void)testTaskForCompletionOfAllTasksWithResultsNoTasksImmediateCompletion {
...
+
+
+- (void)testTasksForTaskForCompletionOfAnyTasksWithErrorAndSuccess {
+    NSError *error = [NSError errorWithDomain:@"BoltsTests"
+                                         code:35
+                                     userInfo:nil];
+    
+    BFTask * task = [BFTask taskForCompletionOfAnyTask:@[[BFTask taskWithError:error], [BFTask taskWithResult:@"success"]]];
```
Comment (nlutsenko) [inline BoltsTests/TaskTests.m]: Nit: Remove space before `task`.

File: BoltsTests/TaskTests.m [hunk 16]
```diff
@@ -628,6 +629,83 @@ - (void)testTaskForCompletionOfAllTasksWithResultsNoTasksImmediateCompletion {
...
+- (void)testTasksForTaskForCompletionOfAnyTasksWithRacing {
+    BFTask *first = [[BFTask taskWithDelay:2] continueWithBlock:^id _Nullable(BFTask * _Nonnull task) {
+        return [BFTask taskWithResult:@"first"];
+    }];
+    BFTask *second = [[BFTask taskWithDelay:3] continueWithBlock:^id _Nullable(BFTask * _Nonnull task) {
+        return [BFTask taskWithResult:@"second"];
+    }];
+
+    
+    BFTask * task = [BFTask taskForCompletionOfAnyTask:@[first, second]];
```
Comment (nlutsenko) [inline BoltsTests/TaskTests.m]: Nit: Remove space before `task`.

File: BoltsTests/TaskTests.m [hunk 17]
```diff
@@ -628,6 +629,83 @@ - (void)testTaskForCompletionOfAllTasksWithResultsNoTasksImmediateCompletion {
...
+    XCTAssertEqualObjects(error, task.error);
+    XCTAssertNotNil(task.error);
+}
+
+- (void)testTasksForTaskForCompletionOfAnyTasksWithNilArray {
+    NSError *error = [NSError errorWithDomain:@"BoltsTests"
+                                         code:35
+                                     userInfo:nil];
+    
+    BFTask * task = [BFTask taskForCompletionOfAnyTask:nil];
```
Comment (nlutsenko) [inline BoltsTests/TaskTests.m]: Nit: Remove space before `task`.

File: BoltsTests/TaskTests.m [hunk 18]
```diff
@@ -628,6 +629,83 @@ - (void)testTaskForCompletionOfAllTasksWithResultsNoTasksImmediateCompletion {
...
+    XCTAssertEqualObjects(@"success", task.result);
+    XCTAssertNil(task.error);
+}
+
+- (void)testTasksForTaskForCompletionOfAnyTasksWithError {
+    NSError *error = [NSError errorWithDomain:@"BoltsTests"
+                                         code:35
+                                     userInfo:nil];
+    
+    BFTask * task = [BFTask taskForCompletionOfAnyTask:@[[BFTask taskWithError:error]]];
```
Comment (nlutsenko) [inline BoltsTests/TaskTests.m]: Nit: Remove space before `task`.

File: BoltsTests/TaskTests.m [hunk 19]
```diff
@@ -628,6 +629,83 @@ - (void)testTaskForCompletionOfAllTasksWithResultsNoTasksImmediateCompletion {
...
+    XCTAssertNil(task.error);
+    XCTAssertNil(task.exception);
+}
+
+- (void)testTasksForTaskForCompletionOfAnyTasksAllErrors {
+    NSError *error = [NSError errorWithDomain:@"BoltsTests"
+                                         code:35
+                                     userInfo:nil];
+    
+    BFTask * task = [BFTask taskForCompletionOfAnyTask:@[[BFTask taskWithError:error], [BFTask taskWithError:error]]];
```
Comment (nlutsenko) [inline BoltsTests/TaskTests.m]: Nit: Remove space before `task`.

File: Vendor/xctoolchain [hunk 20]
```diff
@@ -1 +1 @@
-Subproject commit f1dd6b981643e1bbf5815093d37818d7f8e8e96a
+Subproject commit 94b6a06c0c649d71a11a4d2f7ace28bc8c865acb
```
Comment (nlutsenko) [inline Vendor/xctoolchain]: Revert this change, since I believe this change is accidental (it's a commit from 4 months ago).

**Comment by nlutsenko** [comment]
Almost there. Few nits here and there and almost ready to go.
Also, you might want to use generics for those `NSMutableArray`s on the stack, since it will enforce that you can only add `NSError`/`NSException` to them.
